### PR TITLE
PINT-493 - Add product options to headless PDP buttons

### PR DESCRIPTION
**Pull request by corywebb**
# Description

Update the shortcode, widget, and Gutenberg block for the PDP button to include the `product_options` parameter. Also, update the Gutenberg PDP button block to enable searching for and selecting products and variations instead of having to enter product and variation ID's.

See https://fastaf.atlassian.net/browse/PINT-493

# Testing instructions

1. Login to the [Fast WC dev admin](https://fasttestdev.wpcomstaging.com/wp-admin).
2. Go to Posts or Pages.
3. Add a new post or page.
4. In the Gutenberg editor, click the `+` button to add a block.
5. Search for the Fast Checkout Product Button block.
6. In the right sidebar under Product Details, search for a product to add. (Note, the "Hoodie" and "Beanie" products have variations that you can test.)
7. Select the product you want to use.
8. If the product has variations, select the variation you want to use.
9. If the variation has selectable Product Options, select the product options you want to use.
10. Publish the post or page.
11. View the post or page on the front end and verify that the button works as expected.

# Checks
- [x] Code has been formatted with `phpcbf --standard WordPress`
- [x] No new linter warnings from `phpcs --standard WordPress`

@ilkerulutas @brikr this is ready for code review.

### CODE REVIEW

File: assets/src/blocks/components/product-attributes/index.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	selected,
+} ) => {
+	const [attKeys, setAttKeys] = useState([]);
+	const [labels, setLabels] = useState([]);
+	const [options, setOptions] = useState([]);
+
+	useEffect(
+		() => {
+			getProductAttributes( product, variant )
+				.then( ( list ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product-attributes/index.js:27]: can we use async/await in here? we prefer that in our other frontend code

you might get complaints for having an async function in the useEffect. what we usually do there is wrap the async stuff in a regular function that we immediately call, e.g.
```js
useEffect(() => {
  (async () => {
    await thing;
  })();
});
```

File: assets/src/blocks/components/product-attributes/index.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+					const newOptions = {};
+					const newAttributes = {};
+
+					Object.keys( selected ).map( ( selectedKey ) => {
+						let availableValues = [];
+
+						if ( list.options && list.options[ selectedKey ] && list.options[ selectedKey ].length ) {
```
Comment (brikr) [inline assets/src/blocks/components/product-attributes/index.js:34]: should be able to use optional chaining here
```suggestion
						if ( list.options?.[ selectedKey ]?.length ) {
```

File: assets/src/blocks/components/product-attributes/index.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+							availableValues = list.options[ selectedKey ].map( ( option ) => option.value );
+						}
+
+						if ( list.attKeys && list.attKeys.includes( selectedKey ) && availableValues.includes( selected[ selectedKey ] ) ) {
```
Comment (brikr) [inline assets/src/blocks/components/product-attributes/index.js:38]: more chaining
```suggestion
						if ( list.attKeys?.includes( selectedKey ) && availableValues.includes( selected[ selectedKey ] ) ) {
```

File: assets/src/blocks/components/product-attributes/index.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+							newAttributes[ selectedKey ] = selected[ selectedKey ];
+						}
+					} );
+
+					if ( list.attKeys && list.attKeys.length ) {
```
Comment (brikr) [inline assets/src/blocks/components/product-attributes/index.js:43]:
```suggestion
					if ( list.attKeys?.length ) {
```

File: assets/src/blocks/components/product-attributes/index.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+					setOptions( newOptions );
+				} );
+		},
+		[
+			product,
+			variant,
+		]
+	);
+
+	const onAttributeChange = ( attKey, value ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product-attributes/index.js:78]: nit: to match our react style: we have all props for these named `onX` and the actual functions be called `handleX` to disambiguate which is a prop and which is the actual handler
```suggestion
	const handleAttributeChange = ( attKey, value ) => {
```

File: assets/src/blocks/components/product-variant/index.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+			value: '',
+		},
+	];
+
+	const [options, setOptions] = useState( defaultOptions );
+
+	useEffect(
+		() => {
+			getProductVariations( product )
+				.then( ( list ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product-variant/index.js:40]: use async/await if we can

File: assets/src/blocks/components/product-variant/index.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+					let options = [];
+					let variant_ids = [];
```
Comment (brikr) [inline assets/src/blocks/components/product-variant/index.js:42]: nit: use camelCase for this

File: assets/src/blocks/components/product-variant/index.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+						onChange( '' );
+					}
+
+					setOptions( options );
+				} );
+		},
+		[ product ]
+	);
+
+	const onSelect = ( variant ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product-variant/index.js:65]: same nit as other file
```suggestion
	const handleSelect = ( variant ) => {
```

File: assets/src/blocks/components/product/index.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	const [product, setProduct] = useState( selected );
+
+	useEffect(
+		() => {
+			fetchProducts( { selected } );
+		},
+		[] // Dependencies. Leave blank to avoid running more than once.
+	);
+
+	const onSearch = ( search ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product/index.js:35]:
```suggestion
	const handleSearch = ( search ) => {
```

File: assets/src/blocks/components/product/index.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		setIsLoading( true );
+		fetchProducts( { selected: product, search } );
+	};
+
+	const fetchProducts = ( {
+		selected,
+		search = '',
+	} ) => {
+		getProducts( { selected, search } )
+			.then( ( list ) => {
```
Comment (brikr) [inline assets/src/blocks/components/product/index.js:45]: async/await

File: assets/src/blocks/components/utilities/index.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+/**
+ * Get a promise that resolves to a list of products from the WooCommerce Store API.
+ *
+ * @param {Object} request          A query object with the list of selected products and search term.
+ * @param {Array}  request.selected Currently selected products.
+ * @param {string} request.search   Search string.
+ */
+export const getProducts = (
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:31]: convert to async/await

File: assets/src/blocks/components/utilities/index.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+				}
+			)
+		);
+	}
+
+	requests.push( addQueryArgs( productsBase, queryArgs ) );
+
+	return Promise.all( requests.map( ( path ) => apiFetch( { path } ) ) )
+		.then( ( data ) => {
+			const products = uniqBy( flatten( data ), 'id' );
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:68]: should just be able to use the built-in `flat` function here
```suggestion
			const products = uniqBy( data.flat(), 'id' );
```

File: assets/src/blocks/components/utilities/index.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+import { flatten, uniqBy } from 'lodash';
+
+const apiFetch = wp.apiFetch;
+const { addQueryArgs } = wp.url;
+
+const productsBase = '/wc/store/products';
+
+/**
+ * Get the number of products in the WooCommerce store from the global wcSettings object.
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:14]: can you add a `@returns` jsdoc?

File: assets/src/blocks/components/utilities/index.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+const getProductCount = () => {
+	const wcSettings = window.wcSettings || {};
+	const blocksConfig = wcSettings.hasOwnProperty( 'wcBlocksConfig' ) ? wcSettings.wcBlocksConfig : {};
+	const productCount = blocksConfig.hasOwnProperty( 'productCount' ) ? blocksConfig.productCount : 100;
+
+	return productCount;
+};
+
+/**
+ * Get a promise that resolves to a list of products from the WooCommerce Store API.
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:25]: can you add a `@returns` jsdoc?

File: assets/src/blocks/components/utilities/index.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+			} ) );
+			return list;
+		} )
+		.catch ( ( e ) => {
+			throw e;
+		} );
+};
+
+/**
+ * Get a promise that resolves to a product object from the WooCommerce Store API.
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:81]: can you add a `@returns` jsdoc?

File: assets/src/blocks/components/utilities/index.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+ * @param {number} productId Id of the product to retrieve.
+ */
+export const getProduct = ( productId ) => {
+	return apiFetch( {
+		path: `${ productsBase }/${ productId }`,
+	} );
+};
+
+/**
+ * Get the list of non-variation attributes from a product retrieved using the WooCommerce Store API.
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:92]: can you add a `@returns` jsdoc?

File: assets/src/blocks/components/utilities/index.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+			variantId: variantId,
+		}
+	);
+	const attributes = await apiFetch( { path } );
+
+	return attributes;
+};
+
+/**
+ * Get a promise that resolves to a list of variation objects from the WooCommerce Store API.
```
Comment (brikr) [inline assets/src/blocks/components/utilities/index.js:111]: can you add a `@returns` jsdoc?

**Review by brikr — CHANGES_REQUESTED** [review]
nice work! most of the comments are to match our react code style that we use in monorepo

File: includes/routes.php [hunk 18]
```diff
@@ -107,6 +122,24 @@ function fastwc_api_permission_callback() {
 	return $has_permission;
 }
 
+/**
+ * REST API permissions callback for product attributes
+ *
+ * @return bool
+ */
+function fastwc_api_product_attributes_permission_callback() {
```
Comment (ilkerulutas) [inline includes/routes.php:130]: Question: Would it make sense to generalize the name of this function for reuse in other cases when needed.
Something like fastwc_api_managewc_permission_callback ?

**Review by brikr — COMMENTED** [review]
leaving a comment so github knows i've viewed these files

File: assets/src/blocks/components/product-attributes/index.js (hunk 1 above)
Comment (corywebb) [reply assets/src/blocks/components/product-attributes/index.js]: This has now been updated. I created a named function to call from within `useEffect` and switched to use async/await.

File: includes/routes.php (hunk 18 above)
Comment (corywebb) [reply includes/routes.php:130]: Good idea. This is updated.

**Comment by corywebb** [comment]
@brikr this is ready for re-review.

**Review by brikr — APPROVED** [review]
lgtm!

**Pull request merged**